### PR TITLE
feat: refine data push/pull

### PR DESCRIPTION
**Pull request by romic7x**

### CODE REVIEW

File: maro/cli/grass/utils/copy.py [hunk 1]
```diff
@@ -5,14 +5,15 @@
...
 from maro.cli.utils.subprocess import SubProcess
 from maro.utils.exception.cli_exception import CliException
 from maro.utils.logger import CliLogger
 
 logger = CliLogger(name=__name__)
 
 
-def copy_files_to_node(local_path: str, remote_dir: str, admin_username: str, node_ip_address: str):
+def copy_files_to_node(local_path: str, remote_dir: str, admin_username: str, node_ip_address: str) -> None:
     """Copy local files to node, automatically create folder if not exist
```
Comment (Jinyu-W) [inline maro/cli/grass/utils/copy.py:17]: End with a period.

File: maro/cli/grass/utils/copy.py [hunk 2]
```diff
@@ -21,12 +22,19 @@ def copy_files_to_node(local_path: str, remote_dir: str, admin_username: str, no
...
-                   f"-az {local_path} {admin_username}@{node_ip_address}:{remote_dir}"
-    _ = SubProcess.run(copy_scripts)
+    source_path = get_reformatted_source_path(local_path)
+    basename = os.path.basename(source_path)
+    folder_name = os.path.dirname(source_path)
+    target_dir = get_reformatted_target_dir(remote_dir)
 
+    mkdir_script = f"ssh -o StrictHostKeyChecking=no {admin_username}@{node_ip_address} 'mkdir -p {target_dir}'"
+    _ = SubProcess.run(mkdir_script)
+    copy_script = f"tar czf - -C {folder_name} {basename} | " \
```
Comment (Jinyu-W) [inline maro/cli/grass/utils/copy.py:32]: https://google.github.io/styleguide/pyguide.html#32-line-length

Comment (Jinyu-W) [reply maro/cli/grass/utils/copy.py:32]: copy_script = (f"..."
                        f"...")

File: maro/cli/grass/utils/copy.py [hunk 3]
```diff
@@ -21,12 +22,19 @@ def copy_files_to_node(local_path: str, remote_dir: str, admin_username: str, no
...
+                  f"ssh {admin_username}@{node_ip_address} 'tar xzf - -C {target_dir}'"
+    _ = SubProcess.run(copy_script)
 
-def copy_files_from_node(local_dir: str, remote_path: str, admin_username: str, node_ip_address: str):
+
+def copy_files_from_node(local_dir: str, remote_path: str, admin_username: str, node_ip_address: str) -> None:
     """Copy node files to local, automatically create folder if not exist
```
Comment (Jinyu-W) [inline maro/cli/grass/utils/copy.py:38]: period

File: maro/cli/grass/utils/copy.py [hunk 4]
```diff
@@ -35,9 +43,16 @@ def copy_files_from_node(local_dir: str, remote_path: str, admin_username: str,
...
-                   f"-az {admin_username}@{node_ip_address}:{remote_path} {local_dir}"
-    _ = SubProcess.run(copy_scripts)
+    source_path = get_reformatted_source_path(remote_path)
+    basename = os.path.basename(source_path)
+    folder_name = os.path.dirname(source_path)
+    target_dir = get_reformatted_target_dir(local_dir)
+
+    mkdir_script = f"mkdir -p {target_dir}"
+    _ = SubProcess.run(mkdir_script)
+    copy_script = f"ssh {admin_username}@{node_ip_address} 'tar czf - -C {folder_name} {basename}' | " \
```
Comment (Jinyu-W) [inline maro/cli/grass/utils/copy.py:53]: same as before

File: maro/cli/k8s/executors/k8s_azure_executor.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 import yaml
 
+from maro.cli.utils.copy import get_reformatted_source_path, get_reformatted_target_dir
 from maro.cli.utils.details import load_cluster_details, save_cluster_details, load_job_details, save_job_details, \
     load_schedule_details, save_schedule_details
```
Comment (Jinyu-W) [inline maro/cli/k8s/executors/k8s_azure_executor.py:14]: from ... import (...
...)

instead?

File: maro/cli/utils/copy.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+def get_reformatted_source_path(path: str):
+    """Build source path without trailing '/'.
+
+    Args:
+        path (str): original path.
```
Comment (Jinyu-W) [inline maro/cli/utils/copy.py:5]: Capital

File: maro/cli/utils/copy.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    Returns:
+        str: reformatted path.
```
Comment (Jinyu-W) [inline maro/cli/utils/copy.py:8]: Capital

**Review by Jinyu-W — CHANGES_REQUESTED** [review]

**Review by Jinyu-W — DISMISSED** [review]
* update with the base v0.1
* check the test status of this PR

File: maro/cli/grass/utils/copy.py [hunk 8]
```diff
@@ -5,43 +5,58 @@
...
+    mkdir_script = f"mkdir -p {target_dir}"
+    _ = SubProcess.run(mkdir_script)
+    copy_script = (f"ssh {admin_username}@{node_ip_address} 'tar czf - -C {folder_name} {basename}' | "
+                   f"tar xzf - -C {target_dir}")
+    _ = SubProcess.run(copy_script)
 
 
 def sync_mkdir(remote_path: str, admin_username: str, node_ip_address: str):
-    """mkdir synchronously at local and remote
+    """mkdir synchronously at local and remote.
```
Comment (ArthurJiang) [inline maro/cli/grass/utils/copy.py:59]: Mkdir or make directory

File: tests/cli/config.yml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+cloud/subscription: ""
+user/admin_public_key: ""
```
Comment (ArthurJiang) [inline tests/cli/config.yml:2]: new line

File: tests/cli/grass/test_data/README.md [hunk 10]
```diff
@@ -0,0 +1 @@
+(empty)
```
Comment (ArthurJiang) [inline tests/cli/grass/test_data/README.md:1]: Use .gitkeep for keeping the empty directory

https://stackoverflow.com/questions/7229885/what-are-the-differences-between-gitignore-and-gitkeep

File: tests/cli/grass/test_grass.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+from maro.cli.utils.params import GlobalParams, GlobalPaths
+from maro.cli.utils.subprocess import SubProcess
+
+
+@unittest.skipUnless(os.environ.get("test_with_cli", False), "Require cli prerequisites.")
+class TestGrass(unittest.TestCase):
+    """Tests for Grass Mode.
+
+    Tests should be executed in specific order,
```
Comment (ArthurJiang) [inline tests/cli/grass/test_grass.py:22]:
```python
import unittest
unittest.TestLoader.sortTestMethodsUsing = None
```
This one may also work for this case.

ref: https://stackoverflow.com/questions/4095319/unittest-tests-order/22317851#22317851

File: tests/cli/k8s/test_data/README.md [hunk 12]
```diff
@@ -0,0 +1 @@
+(empty)
```
Comment (ArthurJiang) [inline tests/cli/k8s/test_data/README.md:1]: https://github.com/microsoft/maro/pull/138/files#r501555895

File: tests/cli/grass/test_data/README.md (hunk 10 above)
Comment (romic7x) [reply tests/cli/grass/test_data/README.md:1]: actually, I just want to mock a file here

Comment (romic7x) [reply tests/cli/grass/test_data/README.md:1]: will remove it after the training test is integrated

File: tests/cli/grass/test_grass.py (hunk 11 above)
Comment (romic7x) [reply tests/cli/grass/test_grass.py:22]: apparently, a better way is to define a sorting function to determine the strategy we want to order, and that will incur some redundant workload. So we are not going to change it at this time.

**Review by ArthurJiang — APPROVED** [review]

**Pull request merged**